presto-main/src/main/java/com/facebook/presto/operator/JoinStatisticsCounter.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.operator.LookupJoinOperators.JoinType;
+
+import java.util.function.Supplier;
+
+import static com.facebook.presto.operator.JoinOperatorInfo.createJoinOperatorInfo;
+import static java.util.Objects.requireNonNull;
+
+public class JoinStatisticsCounter
+        implements Supplier<OperatorInfo>
+{
+    public static final int HISTOGRAM_SIZE = 8;
+
+    private final JoinType joinType;
+    // Logarithmic histogram. Regular histogram (or digest) is too expensive, because of memory manipulations. Also, we don't need their guarantees of precision.
+    // To make it maximally fast by reducing indirections (it will fit in cache L1 anyways) counters are packed in one array.
+    // Layout (here "bucket" is histogram bucket):
+    //      [2*bucket]      count probe positions that produced "bucket" rows on source side,
+    //      [2*bucket + 1]  total count of rows that were produces by probe rows in this bucket.
+    private final long[] logHistogramCounters = new long[HISTOGRAM_SIZE * 2];
+
+    public JoinStatisticsCounter(JoinType joinType)
+    {
+        this.joinType = requireNonNull(joinType, "joinType is null");
+    }
+
+    public void recordProbe(int numSourcePosition)
+    {
+        int bucket;
+        if (numSourcePosition <= 4) {
```

Comment:
...and this is an example - I inlined and you have now clue this what is this number. I made another variant with both increase readability and constant make it easy to change number of individually recorded buckets without worrying about dependencies on aggregated bucket indices.